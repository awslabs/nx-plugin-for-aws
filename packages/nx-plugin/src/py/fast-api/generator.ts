/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { FastApiProjectGeneratorSchema } from './schema';
import { UVProvider } from '@nxlv/python/src/provider/uv/provider';
import { Logger } from '@nxlv/python/src/executors/utils/logger';
import pyProjectGenerator, { getPyProjectDetails } from '../project/generator';
import { parse, stringify } from '@iarna/toml';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import { toClassName, toKebabCase, toSnakeCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { addHttpApi } from '../../utils/http-api';
import { sortProjectTargets } from '../../utils/nx';

/**
 * Generates a Python FastAPI project
 */
export const fastApiProjectGenerator = async (
  tree: Tree,
  schema: FastApiProjectGeneratorSchema,
): Promise<GeneratorCallback> => {
  await sharedConstructsGenerator(tree);

  const { dir, normalizedName, normalizedModuleName, fullyQualifiedName } = getPyProjectDetails(
    tree,
    {
      name: schema.name,
      directory: schema.directory,
    },
  );
  const apiNameSnakeCase = toSnakeCase(schema.name);
  const apiNameKebabCase = toKebabCase(schema.name);
  const apiNameClassName = toClassName(schema.name);
  const enhancedOptions = {
    ...schema,
    dir,
    apiNameClassName,
    apiNameKebabCase,
    apiNameSnakeCase,
  };

  await pyProjectGenerator(tree, {
    name: normalizedName,
    directory: schema.directory,
    moduleName: normalizedModuleName,
    projectType: 'application',
  });

  const projectConfig = readProjectConfiguration(tree, fullyQualifiedName);

  projectConfig.targets.bundle = {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/bundle`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${normalizedName} -o dist/${dir}/bundle/requirements.txt`,
        `uv pip install --no-installer-metadata --no-compile-bytecode --python-platform x86_64-manylinux2014 --python \`uv python pin\` --target dist/${dir}/bundle -r dist/${dir}/bundle/requirements.txt`
      ],
      parallel: false,
    },
    dependsOn: ['compile'],
  };
  projectConfig.targets.build.dependsOn = [
    ...(projectConfig.targets.build.dependsOn ?? []),
    'bundle',
  ];
  projectConfig.targets.serve = {
    executor: '@nxlv/python:run-commands',
    options: {
      command: `uv run fastapi dev ${normalizedName}/main.py`,
      cwd: dir,
    },
  };

  projectConfig.targets = sortProjectTargets(projectConfig.targets);
  updateProjectConfiguration(tree, normalizedName, projectConfig);

  [
    joinPathFragments(dir, normalizedModuleName ?? normalizedName, 'hello.py'),
    joinPathFragments(dir, 'tests', 'test_hello.py'),
  ].forEach((f) => tree.delete(f));

  generateFiles(
    tree, // the virtual file system
    joinPathFragments(__dirname, 'files', 'app'), // path to the file templates
    dir, // destination path of the files
    {
      name: normalizedName,
    },
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );

  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'http-apis',
        `${apiNameKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );

    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './http-apis/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'http-apis',
        'index.ts',
      ),
      `./${apiNameKebabCase}.js`,
    );
  }

  addHttpApi(tree, apiNameClassName);

  updateJson(
      tree,
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
      (config: ProjectConfiguration) => {
        if (!config.targets) {
          config.targets = {};
        }
        if (!config.targets.build) {
          config.targets.build = {};
        }
        config.targets.build.dependsOn = [
          ...(config.targets.build.dependsOn ?? []),
          `${fullyQualifiedName}:build`,
        ];
        return config;
      },
    );

  const projectToml = parse(
    tree.read(joinPathFragments(dir, 'pyproject.toml'), 'utf8'),
  ) as UVPyprojectToml;
  projectToml.project.dependencies = ['fastapi', 'mangum'].concat(
    projectToml.project?.dependencies || [],
  );
  projectToml['dependency-groups'] = { dev: ['fastapi[standard]>=0.115'] };
  tree.write(joinPathFragments(dir, 'pyproject.toml'), stringify(projectToml));

  await formatFilesInSubtree(tree);

  return async () => {
    await new UVProvider(tree.root, new Logger(), tree).install();
    installPackagesTask(tree);
  };
};
export default fastApiProjectGenerator;
