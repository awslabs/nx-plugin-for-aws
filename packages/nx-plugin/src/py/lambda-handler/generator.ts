/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { LambdaHandlerProjectGeneratorSchema } from './schema';
import { UVProvider } from '@nxlv/python/src/provider/uv/provider';
import { Logger } from '@nxlv/python/src/executors/utils/logger';
import pyProjectGenerator, { getPyProjectDetails } from '../project/generator';
import { parse, stringify } from '@iarna/toml';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { toClassName, toKebabCase, toSnakeCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { addLambdaHandler } from '../../utils/lambda-handler';
import { sortObjectKeys } from '../../utils/object';
import { NxGeneratorInfo, getGeneratorInfo } from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';

export const LAMBDA_HANDLER_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

console.log('LAMBDA_HANDLER_GENERATOR_INFO', LAMBDA_HANDLER_GENERATOR_INFO);

/**
 * Generates a Python Lambda Handler project
 */
export const lambdaHandlerProjectGenerator = async (
  tree: Tree,
  schema: LambdaHandlerProjectGeneratorSchema,
): Promise<GeneratorCallback> => {
  await sharedConstructsGenerator(tree);

  const { dir, normalizedName, normalizedModuleName, fullyQualifiedName } =
    getPyProjectDetails(tree, {
      name: schema.name,
      directory: schema.directory,
    });

  const lambdaHandlerClassName = toClassName(schema.name);
  const lambdaHandlerKebabCase = toKebabCase(schema.name);
  const lambdaHandlerSnakeCase = toSnakeCase(schema.name);
  const enhancedOptions = {
    ...schema,
    dir,
    lambdaHandlerClassName,
    lambdaHandlerKebabCase,
    lambdaHandlerSnakeCase,
  };

  await pyProjectGenerator(tree, {
    name: normalizedName,
    directory: schema.directory,
    moduleName: normalizedModuleName,
    projectType: 'application',
  });

  const projectConfig = readProjectConfiguration(tree, fullyQualifiedName);

  projectConfig.targets.bundle = {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/bundle`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${normalizedName} -o dist/${dir}/bundle/requirements.txt`,
        `uv pip install -n --no-installer-metadata --no-compile-bytecode --python-platform x86_64-manylinux2014 --target dist/${dir}/bundle -r dist/${dir}/bundle/requirements.txt`,
      ],
      parallel: false,
    },
    dependsOn: ['compile'],
  };
  projectConfig.targets.build.dependsOn = [
    ...(projectConfig.targets.build.dependsOn ?? []),
    'bundle',
  ];

  projectConfig.metadata = {
    handlerName: schema.name,
  } as any;

  projectConfig.targets = sortObjectKeys(projectConfig.targets);
  updateProjectConfiguration(tree, normalizedName, projectConfig);

  [
    joinPathFragments(dir, normalizedModuleName ?? normalizedName, 'hello.py'),
    joinPathFragments(dir, 'tests', 'test_hello.py'),
  ].forEach((f) => tree.delete(f));

  generateFiles(
    tree, // the virtual file system
    joinPathFragments(__dirname, 'files', 'app'), // path to the file templates
    dir, // destination path of the files
    {
      name: normalizedName,
      lambdaHandlerClassName,
    },
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );

  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'lambda-handlers',
        `${lambdaHandlerKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );

    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './lambda-handlers/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'lambda-handlers',
        'index.ts',
      ),
      `./${lambdaHandlerKebabCase}.js`,
    );
  }

  addLambdaHandler(tree, lambdaHandlerKebabCase);

  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []),
        `${fullyQualifiedName}:build`,
      ];
      return config;
    },
  );

  const projectToml = parse(
    tree.read(joinPathFragments(dir, 'pyproject.toml'), 'utf8'),
  ) as UVPyprojectToml;
  projectToml.project.dependencies = [
    'aws-lambda-powertools',
    'aws-lambda-powertools[tracer]',
  ].concat(projectToml.project?.dependencies || []);
  tree.write(joinPathFragments(dir, 'pyproject.toml'), stringify(projectToml));

  await addGeneratorMetricsIfApplicable(tree, [LAMBDA_HANDLER_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);

  return async () => {
    await new UVProvider(tree.root, new Logger(), tree).install();
    installPackagesTask(tree);
  };
};
export default lambdaHandlerProjectGenerator;
