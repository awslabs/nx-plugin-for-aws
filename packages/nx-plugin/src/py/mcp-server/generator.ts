/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  OverwriteStrategy,
  Tree,
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  updateProjectConfiguration,
} from '@nx/devkit';
import { PyMcpServerGeneratorSchema } from './schema';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { formatFilesInSubtree } from '../../utils/format';
import { kebabCase, toClassName, toSnakeCase } from '../../utils/names';
import { addDependenciesToPyProjectToml } from '../../utils/py';
import { getNpmScope } from '../../utils/npm-scope';
import { addMcpServerInfra } from '../../utils/agent-core-constructs/agent-core-constructs';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { addPythonBundleTarget } from '../../utils/bundle/bundle';
import { withVersions } from '../../utils/versions';
import { Logger } from '@nxlv/python/src/executors/utils/logger';
import { UVProvider } from '@nxlv/python/src/provider/uv/provider';
import { resolveIacProvider } from '../../utils/iac';

export const PY_MCP_SERVER_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export const pyMcpServerGenerator = async (
  tree: Tree,
  options: PyMcpServerGeneratorSchema,
): Promise<GeneratorCallback> => {
  const project = readProjectConfigurationUnqualified(tree, options.project);

  const pyProjectPath = joinPathFragments(project.root, 'pyproject.toml');

  // Check if the project has a pyproject.toml file
  if (!pyProjectPath) {
    throw new Error(
      `Unsupported project ${options.project}. Expected a Python project (with a pyproject.toml)`,
    );
  }

  if (!project.sourceRoot) {
    throw new Error(
      `This project does not have a source root. Please add a source root to the project configuration before running this generator.`,
    );
  }

  // Module name is the last part of the source root,
  const sourceParts = project.sourceRoot.split('/');
  const moduleName = sourceParts[sourceParts.length - 1];

  const name = kebabCase(
    options.name ?? `${kebabCase(project.name.split('.').pop())}-mcp-server`,
  );

  const mcpServerNameSnakeCase = toSnakeCase(options.name ?? 'mcp-server');
  const mcpServerNameClassName = toClassName(name);

  const targetSourceDir = joinPathFragments(
    project.sourceRoot,
    mcpServerNameSnakeCase,
  );
  const distDir = joinPathFragments('dist', project.root);

  const computeType = options.computeType ?? 'BedrockAgentCoreRuntime';

  // Generate example server
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    targetSourceDir,
    {
      name,
      mcpServerNameSnakeCase,
      mcpServerNameClassName,
      moduleName,
      distDir,
    },
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  addDependenciesToPyProjectToml(tree, project.root, [
    'mcp',
    'boto3',
    'aws-opentelemetry-distro',
  ]);

  if (computeType === 'BedrockAgentCoreRuntime') {
    const dockerImageTag = `${getNpmScope(tree)}-${name}:latest`;

    // Add bundle target
    addPythonBundleTarget(project, {
      pythonPlatform: 'aarch64-manylinux2014',
    });

    const dockerTargetName = `${name}-docker`;

    // Add a docker target specific to this MCP server
    project.targets[dockerTargetName] = {
      cache: true,
      executor: 'nx:run-commands',
      options: {
        commands: [
          `docker build --platform linux/arm64 -t ${dockerImageTag} ${targetSourceDir} --build-context workspace=.`,
        ],
        parallel: false,
      },
      dependsOn: ['bundle'],
    };

    project.targets.docker = {
      ...project.targets.docker,
      dependsOn: [
        ...(project.targets.docker?.dependsOn ?? []).filter(
          (t) => t !== dockerTargetName,
        ),
        dockerTargetName,
      ],
    };

    project.targets.build = {
      ...project.targets.build,
      dependsOn: [
        ...(project.targets.build?.dependsOn ?? []).filter(
          (t) => t !== 'docker',
        ),
        'docker',
      ],
    };

    // Add shared constructs
    const iacProvider = await resolveIacProvider(tree, options.iacProvider);
    await sharedConstructsGenerator(tree, { iacProvider });

    // Add the construct to deploy the mcp server
    addMcpServerInfra(tree, {
      mcpServerNameKebabCase: name,
      mcpServerNameClassName,
      projectName: project.name,
      dockerImageTag,
      iacProvider,
    });
  } else {
    // No Dockerfile needed for non-hosted MCP
    tree.delete(joinPathFragments(targetSourceDir, 'Dockerfile'));
  }

  updateProjectConfiguration(tree, project.name, {
    ...project,
    targets: {
      ...project.targets,
      // Add targets for running the MCP server
      [`${options.name ? name : 'mcp-server'}-serve-stdio`]: {
        executor: 'nx:run-commands',
        options: {
          commands: [`uv run -m ${moduleName}.${mcpServerNameSnakeCase}.stdio`],
          cwd: '{projectRoot}',
        },
      },
      [`${options.name ? name : 'mcp-server'}-serve-http`]: {
        executor: 'nx:run-commands',
        options: {
          commands: [`uv run -m ${moduleName}.${mcpServerNameSnakeCase}.http`],
          cwd: '{projectRoot}',
        },
      },
      [`${options.name ? name : 'mcp-server'}-inspect`]: {
        executor: 'nx:run-commands',
        options: {
          commands: [
            `mcp-inspector -- uv run -m ${moduleName}.${mcpServerNameSnakeCase}.stdio`,
          ],
          cwd: '{projectRoot}',
        },
      },
    },
  });

  addDependenciesToPackageJson(
    tree,
    {},
    withVersions(['@modelcontextprotocol/inspector']),
  );

  await addGeneratorMetricsIfApplicable(tree, [PY_MCP_SERVER_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return async () => {
    installPackagesTask(tree);
    await new UVProvider(tree.root, new Logger(), tree).install();
  };
};

export default pyMcpServerGenerator;
