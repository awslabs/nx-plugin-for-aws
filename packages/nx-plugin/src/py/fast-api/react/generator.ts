/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { FastApiReactGeneratorSchema } from './schema';
import { runtimeConfigGenerator } from '../../../cloudscape-website/runtime-config/generator';
import snakeCase from 'lodash.snakecase';
import * as path from 'path';
import kebabCase from 'lodash.kebabcase';
import { sortObjectKeys } from '../../../utils/nx';
import { toClassName } from '../../../utils/names';
import { formatFilesInSubtree } from '../../../utils/format';
import { withVersions } from '../../../utils/versions';
import { updateGitIgnore } from '../../../utils/git';

export const fastApiReactGenerator = async (
  tree: Tree,
  options: FastApiReactGeneratorSchema,
) => {
  const frontendProjectConfig = readProjectConfiguration(
    tree,
    options.frontendProjectName,
  );
  const fastApiProjectConfig = readProjectConfiguration(
    tree,
    options.fastApiProjectName,
  );
  const moduleName = getFastApiModuleName(fastApiProjectConfig);

  // Add OpenAPI spec generation script to FastAPI spec (if it does not exist already)
  generateFiles(
    tree,
    path.join(__dirname, 'files/fast-api'),
    fastApiProjectConfig.root,
    {
      moduleName,
    },
  );

  // Instrument the script as part of the fastapi project build
  const fastApiOpenApiDist = joinPathFragments(
    'dist',
    fastApiProjectConfig.root,
    'openapi',
  );
  const specPath = joinPathFragments(fastApiOpenApiDist, 'openapi.json');
  updateProjectConfiguration(tree, options.fastApiProjectName, {
    ...fastApiProjectConfig,
    targets: sortObjectKeys({
      ...fastApiProjectConfig.targets,
      build: {
        ...fastApiProjectConfig.targets?.build,
        dependsOn: [
          ...(fastApiProjectConfig.targets?.build?.dependsOn ?? []).filter(
            (t) => t !== 'openapi',
          ),
          'openapi',
        ],
      },
      openapi: {
        cache: true,
        executor: 'nx:run-commands',
        outputs: [joinPathFragments('{workspaceRoot}', fastApiOpenApiDist)],
        options: {
          commands: [
            `uv run python ${joinPathFragments(fastApiProjectConfig.root, 'scripts', 'generate_open_api.py')} "${specPath}"`,
          ],
        },
      },
    }),
  });

  const apiName = (fastApiProjectConfig.metadata as any)?.apiName;
  const clientGenTarget = `generate:${kebabCase(apiName)}-client`;

  const generatedClientDir = joinPathFragments('generated', kebabCase(apiName));
  const generatedClientDirFromRoot = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    generatedClientDir,
  );

  // Add TypeScript client generation to Frontend project.json
  updateProjectConfiguration(tree, options.frontendProjectName, {
    ...frontendProjectConfig,
    targets: sortObjectKeys({
      ...frontendProjectConfig.targets,
      // Generate should run before compile as the client is created as part of the website src
      compile: {
        ...frontendProjectConfig.targets?.compile,
        dependsOn: [
          ...(frontendProjectConfig.targets?.compile?.dependsOn ?? []).filter(
            (t) => t !== clientGenTarget,
          ),
          clientGenTarget,
        ],
      },
      [clientGenTarget]: {
        cache: true,
        executor: 'nx:run-commands',
        inputs: [
          {
            dependentTasksOutputFiles: '**/*.json',
          },
        ],
        outputs: [
          joinPathFragments('{workspaceRoot}', generatedClientDirFromRoot),
        ],
        options: {
          commands: [
            `nx g @aws/nx-plugin:open-api#ts-client --openApiSpecPath="${specPath}" --outputPath="${generatedClientDirFromRoot}" --no-interactive`,
          ],
        },
        dependsOn: [`${options.fastApiProjectName}:openapi`],
      },
    }),
  });

  const relativeSrcDir = frontendProjectConfig.sourceRoot.slice(
    frontendProjectConfig.root.length + 1,
  );

  // Ignore the generated client by default
  // Users can safely remove the entry from the .gitignore if they prefer to check it in
  updateGitIgnore(tree, frontendProjectConfig.root, (patterns) => [
    ...patterns,
    joinPathFragments(relativeSrcDir, generatedClientDir),
  ]);

  // Ensure that the frontend has runtime config as we'll use the url for creating the client
  await runtimeConfigGenerator(tree, {
    project: options.frontendProjectName,
  });

  // Add sigv4 fetch
  if (options.auth === 'IAM') {
    generateFiles(
      tree,
      joinPathFragments(__dirname, '../../../utils/files/website/hooks/sigv4'),
      joinPathFragments(frontendProjectConfig.sourceRoot, 'hooks'),
      {},
    );
  }

  // Add a hook to instantiate the client
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'website'),
    frontendProjectConfig.sourceRoot,
    {
      auth: options.auth,
      apiName,
      apiNameClassName: toClassName(apiName),
      generatedClientDir,
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      ...((options.auth === 'IAM'
        ? [
            'oidc-client-ts',
            'react-oidc-context',
            '@aws-sdk/client-cognito-identity',
            '@aws-sdk/credential-provider-cognito-identity',
            'aws4fetch',
          ]
        : []) as any),
    ]),
    withVersions(['@smithy/types']),
  );

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

const getFastApiModuleName = (projectConfig: ProjectConfiguration): string => {
  if (projectConfig.sourceRoot) {
    const sourceRootParts = projectConfig.sourceRoot.split('/');
    return sourceRootParts[sourceRootParts.length - 1];
  }
  const apiName = (projectConfig.metadata as any)?.apiName;
  if (apiName) {
    return snakeCase(apiName);
  }
  new Error(`Could not determine sourceRoot for project ${projectConfig.name}`);
};

export default fastApiReactGenerator;
