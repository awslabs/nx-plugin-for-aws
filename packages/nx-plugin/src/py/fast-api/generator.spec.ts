/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { fastApiProjectGenerator } from './generator';
import { parse } from '@iarna/toml';
import { PACKAGES_DIR, SHARED_CONSTRUCTS_DIR } from '../../utils/shared-constructs';
import { joinPathFragments } from '@nx/devkit';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';

describe('fastapi project generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate a FastAPI project with correct structure', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    // Verify project structure
    expect(tree.exists('apps/test_api')).toBeTruthy();
    expect(tree.exists('apps/test_api/test_api')).toBeTruthy();
    expect(tree.exists('apps/test_api/test_api/main.py')).toBeTruthy();
    expect(tree.exists('apps/test_api/tests/test_main.py')).toBeTruthy();

    // Verify default files are removed
    expect(tree.exists('apps/test_api/test_api/hello.py')).toBeFalsy();
    expect(tree.exists('apps/test_api/tests/test_hello.py')).toBeFalsy();
  });

  it('should set up project configuration with FastAPI targets', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    const projectConfig = JSON.parse(
      tree.read('apps/test_api/project.json', 'utf-8')
    );

    // Verify FastAPI-specific targets
    expect(projectConfig.targets.bundle).toBeDefined();
    expect(projectConfig.targets.bundle.outputs).toEqual([
      '{workspaceRoot}/dist/apps/test_api/bundle',
    ]);
    expect(projectConfig.targets.bundle.options.commands).toContain(
      'uv export --frozen --no-dev --no-editable --project test_api -o dist/apps/test_api/bundle/requirements.txt'
    );

    // Verify start target for development
    expect(projectConfig.targets.serve).toBeDefined();
    expect(projectConfig.targets.serve.executor).toBe('@nxlv/python:run-commands');
    expect(projectConfig.targets.serve.options.command).toBe(
      'uv run fastapi dev test_api/main.py'
    );

    // Verify build dependencies
    expect(projectConfig.targets.build.dependsOn).toContain('bundle');
  });

  it('should configure FastAPI dependencies', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    const pyprojectToml = parse(
      tree.read('apps/test_api/pyproject.toml', 'utf-8')
    ) as UVPyprojectToml;

    // Verify FastAPI dependencies
    expect(pyprojectToml.project.dependencies).toContain('fastapi');
    expect(pyprojectToml.project.dependencies).toContain('mangum');
    expect(pyprojectToml['dependency-groups'].dev).toContain('fastapi[standard]>=0.115');
  });

  it('should set up shared constructs for HTTP API', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    // Verify shared constructs files
    const httpApiPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'http-apis',
      'test-api.ts'
    );
    expect(tree.exists(httpApiPath)).toBeTruthy();

    // Verify exports in index files
    const httpApisIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'http-apis',
      'index.ts'
    );
    expect(tree.read(httpApisIndexPath, 'utf-8')).toContain(
      './test-api.js'
    );

    const appIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts'
    );
    expect(tree.read(appIndexPath, 'utf-8')).toContain(
      './http-apis/index.js'
    );
  });

  it('should update shared constructs build dependencies', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    const sharedConstructsConfig = JSON.parse(
      tree.read(
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
        'utf-8'
      )
    );

    expect(sharedConstructsConfig.targets.build.dependsOn).toContain(
      'proj.test_api:build'
    );
  });

  it('should handle custom directory path', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps/nested/path',
    });

    expect(tree.exists('apps/nested/path/test_api')).toBeTruthy();
    expect(tree.exists('apps/nested/path/test_api/test_api/main.py')).toBeTruthy();
  });

  it('should generate HTTP API construct with correct class name', async () => {
    await fastApiProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apps',
    });

    const httpApiPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'http-apis',
      'test-api.ts'
    );
    const httpApiContent = tree.read(httpApiPath, 'utf-8');

    expect(httpApiContent).toContain('export class TestApi extends HttpApi');
  });
});
