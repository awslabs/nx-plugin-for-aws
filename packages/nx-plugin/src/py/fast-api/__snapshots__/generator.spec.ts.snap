// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fastapi project generator > should match snapshot > main-snapshot 1`] = `
{
  "apps/test_api/test_api/__init__.py": """"Automatically generated by Nx."""
",
  "apps/test_api/test_api/main.py": "from fastapi import FastAPI
from mangum import Mangum

app = FastAPI(
    title="TestApi"
)
handler = Mangum(app)

@app.get("/")
def read_root():
    return {"Hello": "World"}",
  "apps/test_api/tests/__init__.py": """"unit tests."""
",
  "apps/test_api/tests/conftest.py": """"Unit tests configuration module."""

pytest_plugins = []
",
  "apps/test_api/tests/test_main.py": "# 
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
# 

def test_main():
    pass
",
}
`;
